File: VB/SchedulerMappedLabels/CustomControl/CustomSchedulerControl.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.XtraScheduler

Namespace SchedulerMappedLabels

    Public Class CustomSchedulerControl
        Inherits SchedulerControl

        Private labelsDataSourceField As Object

        Const defaultLabelIdMappedName As String = "Id"

        Const defaultLabelColorMappedName As String = "Color"

        Const defaultLabelDisplayNameMappedName As String = "DisplayName"

        Public Sub New()
            LabelIdMappedName = defaultLabelIdMappedName
            LabelColorMappedName = defaultLabelColorMappedName
            LabelDisplayNameMappedName = defaultLabelDisplayNameMappedName
        End Sub

        <AttributeProvider(GetType(IListSource))>
        <Category("Data")>
        Public Property LabelsDataSource As Object
            Get
                Return labelsDataSourceField
            End Get

            Set(ByVal value As Object)
                labelsDataSourceField = value
                PopulateLabelsStorage()
                Refresh()
            End Set
        End Property

        <Category("Data")>
        <DefaultValue(defaultLabelIdMappedName)>
        <TypeConverter(GetType(LabelColumnNameConverter))>
        Public Property LabelIdMappedName As String

        <Category("Data")>
        <DefaultValue(defaultLabelColorMappedName)>
        <TypeConverter(GetType(LabelColumnNameConverter))>
        Public Property LabelColorMappedName As String

        <Category("Data")>
        <DefaultValue(defaultLabelDisplayNameMappedName)>
        <TypeConverter(GetType(LabelColumnNameConverter))>
        Public Property LabelDisplayNameMappedName As String

        Public Sub PopulateLabelsStorage()
            Dim labelsController As DataBindingController = New DataBindingController(labelsDataSourceField, String.Empty)
            If Storage Is Nothing Then Return
            Storage.Appointments.Labels.Clear()
            For i As Integer = 0 To labelsController.ItemsCount - 1
                Dim labelId As Object = labelsController.GetRowValue(LabelIdMappedName, i)
                Dim labelDisplayName As String = labelsController.GetRowValue(LabelDisplayNameMappedName, i).ToString()
                Dim labelColor As Color = Color.FromArgb(Convert.ToInt32(labelsController.GetRowValue(LabelColorMappedName, i)))
                Storage.Appointments.Labels.Add(labelId, labelDisplayName, labelDisplayName, labelColor)
            Next
        End Sub
    End Class
End Namespace

File: CS/SchedulerMappedLabels/CustomControl/CustomSchedulerControl.cs
using System;
using System.ComponentModel;
using System.Drawing;
using DevExpress.XtraScheduler;

namespace SchedulerMappedLabels {
    public class CustomSchedulerControl : SchedulerControl {
        object labelsDataSource;
        const string defaultLabelIdMappedName = "Id";
        const string defaultLabelColorMappedName = "Color";
        const string defaultLabelDisplayNameMappedName = "DisplayName";

        public CustomSchedulerControl() {
            LabelIdMappedName = defaultLabelIdMappedName;
            LabelColorMappedName = defaultLabelColorMappedName;
            LabelDisplayNameMappedName = defaultLabelDisplayNameMappedName;
        }

        [AttributeProvider(typeof(IListSource))]
        [Category("Data")]
        public object LabelsDataSource {
            get { return labelsDataSource; }
            set {
                labelsDataSource = value;
                PopulateLabelsStorage();
                Refresh();
            }
        }

        [Category("Data")]
        [DefaultValue(defaultLabelIdMappedName)]
        [TypeConverter(typeof(LabelColumnNameConverter))]
        public string LabelIdMappedName { get; set; }

        [Category("Data")]
        [DefaultValue(defaultLabelColorMappedName)]
        [TypeConverter(typeof(LabelColumnNameConverter))]
        public string LabelColorMappedName { get; set; }

        [Category("Data")]
        [DefaultValue(defaultLabelDisplayNameMappedName)]
        [TypeConverter(typeof(LabelColumnNameConverter))]
        public string LabelDisplayNameMappedName { get; set; }

        public void PopulateLabelsStorage() {
            DataBindingController labelsController = new DataBindingController(labelsDataSource, string.Empty);
            if (Storage == null)
                return;

            Storage.Appointments.Labels.Clear();
            for (int i = 0; i < labelsController.ItemsCount; i++) {
                object labelId = labelsController.GetRowValue(LabelIdMappedName, i);
                string labelDisplayName = labelsController.GetRowValue(LabelDisplayNameMappedName, i).ToString();
                Color labelColor = Color.FromArgb(Convert.ToInt32(labelsController.GetRowValue(LabelColorMappedName, i)));
                Storage.Appointments.Labels.Add(labelId, labelDisplayName, labelDisplayName, labelColor);
            }
        }
    }
}